### PR TITLE
axp - Chapter 9 - Regular Expression References

### PR DESCRIPTION
**Pull request by a-panayotopoulos**
I apologise for my Exercise 1 solution, truly.

### CODE REVIEW

**Comment by thatguy59** [comment]
I will review this in the morning tomorrow - As I'm behind with my solutions I don't want to bias myself starting mine.

File: chapter09/axp/ex01 [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	'There is Mrs. Howell, Ginger, and Gilligan',
+	\%patterns
+);	
+	
+sub rightmost {
+	my ( $string, $patterns ) = @_;
+
+	return (
+		reduce {
+			my $b_idx = $string =~ m/$patterns->{$b}/ ? $-[0] : -1;
```
Comment (thatguy59) [inline chapter09/axp/ex01]: I don't remember coming across `$-[0]` before so if anyone else is in the same boat:

```
$-[0] is the offset of the start of the last successful match.
```

File: chapter09/axp/ex02 [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env perl
+
+use 5.014;
+use warnings;
+
+my $fname = $ARGV[0] // 'patterns.dat';
+my $finish = $^O eq "MSWin32" ? "<Ctrl>+Z" : "<Ctrl>+D";
+
+open my $file, $fname or die "Couldn't read $fname: $!\n";
+my @pats = map { chomp; qr/$_/ } <$file>;
```
Comment (thatguy59) [inline chapter09/axp/ex02]: This is nice and concise, but what happens when there's an invalid pattern?

File: chapter09/axp/ex03 [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+use 5.014;
+use warnings;
+use Regexp::Assemble;
+
+my $fname = $ARGV[0] // 'patterns.dat';
+my $finish = $^O eq "MSWin32" ? "<Ctrl>+Z" : "<Ctrl>+D";
+
+open my $file, $fname or die "Couldn't read $fname: $!\n";
+my $regex = Regexp::Assemble->new->add( <$file> )->re;
```
Comment (thatguy59) [inline chapter09/axp/ex03]: Oooh, I din't know that you could just pass it the file like this.

File: chapter09/axp/ex01 (hunk 1 above)
Comment (alexdeas) [inline chapter09/axp/ex01]: It was in the chapter when talking about the rightmost function. Who thinks @thatguy59 didn't read the chapter?

File: chapter09/axp/ex02 [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+close $file;
+
+say "Patterns";
+say "--------";
+say foreach @pats;
+say "\nNow type lines of text; $finish to finish.";
+
+while ( <> ) {
+	foreach my $pattern ( @pats ) {
+		print "$.\t${^PREMATCH}<<<${^MATCH}>>>${^POSTMATCH}" while ( /$pattern/g );
```
Comment (alexdeas) [inline chapter09/axp/ex02]: These named match variables are useful, but don't use them in the DPG as they have an impact on performance.

File: chapter09/axp/ex02 (hunk 2 above)
Comment (a-panayotopoulos) [inline chapter09/axp/ex02]: Yeah, it crashes and burns :fire:

File: chapter09/axp/ex03 (hunk 3 above)
Comment (a-panayotopoulos) [inline chapter09/axp/ex03]: This is just the good ol' [readline operator](http://perldoc.perl.org/perlop.html#I%2fO-Operators) passing an array to `add`.

File: chapter09/axp/ex02 (hunk 4 above)
Comment (a-panayotopoulos) [inline chapter09/axp/ex02]: I thought it was ``$` ``, `$&` and `$'` that had the performance penalty?

**Pull request merged**